### PR TITLE
Fix url_prefix argument for nesting blueprint

### LINKED ISSUE #4037: The Blueprint(url_prefix) parameter is not working when using nesting blueprint
Minimal example:

```python
from flask import Flask, Blueprint

app = Flask(__name__)
parent = Blueprint('parent', __name__, url_prefix='/child')
child = Blueprint('child', __name__, url_prefix='/parent')


@child.get('/')
def index():
    return 'hello'


parent.register_blueprint(child)
app.register_blueprint(parent)
```

The output of `flask routes`:

```
$ flask routes
Endpoint            Methods  Rule
------------------  -------  -----------------------
parent.child.index  GET      /
static              GET      /static/<path:filename>
```

The URL rule should be `/parent/child/` instead of `/`.

P.S. The URL prefix works fine if set the prefix by using `register_blueprint(url_prefix)`.

Environment:

- Flask version: 2.0

### PR DESCRIPTION
**Pull request by greyli**
fixes #4037

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

File: tests/test_blueprints.py [hunk 1]
```diff
@@ -899,3 +899,51 @@ def grandchild_no():
     assert client.get("/parent/no").data == b"Parent no"
     assert client.get("/parent/child/no").data == b"Parent no"
     assert client.get("/parent/child/grandchild/no").data == b"Grandchild no"
+
+
+def test_nested_blueprint_url_prefix(app, client):
```
Comment (pgjones) [inline tests/test_blueprints.py:904]: This test seems too convoluted for what it tests. Could you simplify it to test the prefix alone - for example using the snippet in #4037?

**Review by pgjones — CHANGES_REQUESTED** [review]

File: tests/test_blueprints.py (hunk 1 above)
Comment (greyli) [reply tests/test_blueprints.py:904]: Sure. Just updated.

**Review by pgjones — APPROVED** [review]

File: src/flask/blueprints.py [hunk 2]
```diff
@@ -340,11 +340,10 @@ def extend(bp_dict, parent_dict):
                 app.cli.add_command(self.cli)
 
         for blueprint, bp_options in self._blueprints:
-            url_prefix = options.get("url_prefix", "")
-            if "url_prefix" in bp_options:
-                url_prefix = (
-                    url_prefix.rstrip("/") + "/" + bp_options["url_prefix"].lstrip("/")
-                )
+            url_prefix = options.get("url_prefix", self.url_prefix)
```
Comment (pgjones) [inline src/flask/blueprints.py:343]: I think this should be,
```python
url_prefix = options.get("url_prefix") or self.url_prefix                                                                                                                                       
bp_url_prefix = bp_options.get("url_prefix") or blueprint.url_prefix                                                                                                                            
if bp_url_prefix and url_prefix:                                                                                                                                                                
    url_prefix = url_prefix.rstrip("/") + "/" + bp_url_prefix.lstrip("/")
```
As it is possible for `url_prefix` to be in the options with a value of `None` and the net url_prefix to also be `None`. What do you think? (This is my [fix](https://gitlab.com/pgjones/quart/-/commit/29e4157583fd10bd504346fc3f319908339acc8d) for Quart).

Comment (davidism) [reply src/flask/blueprints.py:343]: That makes sense, currently rebasing this PR, I'll make that change.

**Comment by davidism** [comment]
@pgjones I noticed while working on this that we're registering blueprints differently in blueprints vs apps. I would expect `Blueprint.register_blueprint` to immediately register all its stuff on the parent blueprint, just like `Flask.register_blueprint` works. I think we should look into refactoring that later.

**Pull request merged**